src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -111,26 +121,268 @@ internal override IEnumerable<TSource> GetEnumerable(int index)
                 // that has its _nextIndex equal to index.  If we don't find one, then it
                 // must be prior to any of them, so call GetEnumerable on the previous
                 // Concat2Iterator.  This avoids a deep recursive call chain.
-                ConcatNIterator<TSource> current = this;
+                ConcatNEnumerableIterator<TSource> current = this;
+                while (true)
+                {
+                    if (index == current._nextIndex)
+                    {
+                        return current._next;
+                    }
+
+                    ConcatIterator<TSource> previous = current._previousConcat;
+
+                    var previousEnumerables = previous as ConcatNEnumerableIterator<TSource>;
+                    if (previousEnumerables != null)
+                    {
+                        current = previousEnumerables;
+                        continue;
+                    }
+
+                    var previousCollections = previous as ConcatNCollectionIterator<TSource>;
+                    if (previousCollections != null)
+                    {
+                        // Since ConcatNCollectionIterator.GetEnumerable does not call into this method,
+                        // it is safe to call GetEnumerable on it here. It also makes things faster, since
+                        // the above type-cast will only ever be run once per call of this method.
```

Comment:
@stephentoub I've added a new commit (7a30a9f) which adds the tests you've suggested.